File: backend/src/modules/dialplan/dialplan.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Dialplan } from './entities/dialplan.entity';
import { PaginationDto, PaginatedResult } from '../../common/dto/pagination.dto';
import { KamailioRpcService } from '../../common/services/kamailio-rpc.service';

@Injectable()
export class DialplanService {
  constructor(
    @InjectRepository(Dialplan)
    private dialplanRepository: Repository<Dialplan>,
    private kamailioRpcService: KamailioRpcService,
  ) {}

  async findAll(dto: PaginationDto & { dpid?: number }): Promise<PaginatedResult<Dialplan>> {
    const { page, limit, dpid } = dto;
    const skip = (page - 1) * limit;
    const where: any = {};
    if (dpid !== undefined) where.dpid = dpid;
    const [items, total] = await this.dialplanRepository.findAndCount({
      where, skip, take: limit, order: { dpid: 'ASC', priority: 'ASC' },
    });
    return new PaginatedResult(items, total, page, limit);
  }

  async create(data: Partial<Dialplan>): Promise<Dialplan> {
    return this.dialplanRepository.save(this.dialplanRepository.create(data));
  }

  async update(id: number, data: Partial<Dialplan>): Promise<Dialplan> {
    const item = await this.dialplanRepository.findOne({ where: { id } });
    if (!item) throw new NotFoundException('规则不存在');
    Object.assign(item, data);
    return this.dialplanRepository.save(item);
  }

  async remove(id: number): Promise<void> {
    await this.dialplanRepository.delete(id);
  }

  async reload(): Promise<void> {
    await this.kamailioRpcService.call('dialplan.reload');
  }

  async translate(dpid: number, input: string): Promise<any> {
    return this.kamailioRpcService.call('dialplan.translate', [dpid, input]);
  }

  async dump(dpid?: number): Promise<any> {
    if (dpid !== undefined) {
      return this.kamailioRpcService.call('dialplan.dump', [dpid]);
    }
    return this.kamailioRpcService.call('dialplan.dump');
  }
}

File: kam-admin-module-routing/src/main/java/io/kamailio/admin/modules/dialplan/service/DialplanService.java
package io.kamailio.admin.modules.dialplan.service;

import io.kamailio.admin.common.RpcTimeouts;
import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.common.dto.PaginationDto;
import io.kamailio.admin.common.service.KamailioRpcService;
import io.kamailio.admin.modules.dialplan.entity.Dialplan;
import io.kamailio.admin.modules.dialplan.repository.DialplanRepository;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class DialplanService {
    private final DialplanRepository repository;
    private final KamailioRpcService kamailioRpc;

    public DialplanService(DialplanRepository repository, KamailioRpcService kamailioRpc) {
        this.repository = repository;
        this.kamailioRpc = kamailioRpc;
    }

    public PaginatedResult<Dialplan> findAll(PaginationDto dto, Integer dpid) {
        var sort = Sort.by("dpid", "priority");
        var pageable = PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), sort);
        var page = dpid != null ? repository.findByDpid(dpid, pageable) : repository.findAll(pageable);
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(), dto.pageOrDefault(), dto.limitOrDefault());
    }

    @Transactional
    public Dialplan create(Dialplan data) { return repository.save(data); }

    @Transactional
    public Dialplan update(Integer id, Dialplan data) {
        var item = repository.findById(id).orElseThrow(() -> new RuntimeException("规则不存在"));
        if (data.getDpid() != null) item.setDpid(data.getDpid());
        if (data.getPriority() != null) item.setPriority(data.getPriority());
        if (data.getMatchOp() != null) item.setMatchOp(data.getMatchOp());
        if (data.getMatchExp() != null) item.setMatchExp(data.getMatchExp());
        if (data.getMatchLen() != null) item.setMatchLen(data.getMatchLen());
        if (data.getSubstExp() != null) item.setSubstExp(data.getSubstExp());
        if (data.getReplExp() != null) item.setReplExp(data.getReplExp());
        if (data.getAttrs() != null) item.setAttrs(data.getAttrs());
        return repository.save(item);
    }

    @Transactional
    public void remove(Integer id) { repository.deleteById(id); }

    public void reload() { kamailioRpc.dialplanReload().block(RpcTimeouts.DEFAULT_BLOCK); }
    public Object translate(int dpid, String input) { return kamailioRpc.dialplanTranslate(dpid, input).block(RpcTimeouts.DEFAULT_BLOCK); }
    public Object dump(Integer dpid) { return kamailioRpc.dialplanDump(dpid).block(RpcTimeouts.DEFAULT_BLOCK); }
}
